node/core/approval-voting/src/approval_checking.rs

```diff
@@ -993,6 +1016,205 @@ mod tests {
+			approvals: vec![1],
+			clock_drift: 10,
+			no_show_duration: 10,
+			drifted_tick_now: 20,
+			exp_no_shows: 0,
+			exp_next_no_show: None,
+		})
+	}
+
+	#[test]
+	fn count_no_shows_two_validators_next_no_show_ordered_first() {
+		test_count_no_shows(NoShowTest {
+			assignments: vec![(ValidatorIndex(1), 21), (ValidatorIndex(2), 22)],
+			approvals: vec![],
+			clock_drift: 10,
+			no_show_duration: 10,
+			drifted_tick_now: 20,
+			exp_no_shows: 0,
+			exp_next_no_show: Some(31),
+		})
+	}
+
+	#[test]
+	fn count_no_shows_two_validators_next_no_show_ordered_last() {
+		test_count_no_shows(NoShowTest {
+			assignments: vec![(ValidatorIndex(1), 22), (ValidatorIndex(2), 21)],
+			approvals: vec![],
+			clock_drift: 10,
+			no_show_duration: 10,
+			drifted_tick_now: 20,
+			exp_no_shows: 0,
+			exp_next_no_show: Some(31),
+		})
+	}
+
+	#[test]
+	fn count_no_shows_three_validators_one_almost_late_one_no_show_one_approving() {
+		test_count_no_shows(NoShowTest {
+			assignments: vec![(ValidatorIndex(1), 21), (ValidatorIndex(2), 20), (ValidatorIndex(3), 20)],
+			approvals: vec![3],
+			clock_drift: 10,
+			no_show_duration: 10,
+			drifted_tick_now: 20,
+			exp_no_shows: 1,
+			exp_next_no_show: Some(31),
+		})
+	}
+
+	#[test]
+	fn count_no_shows_three_no_show_validators() {
+		test_count_no_shows(NoShowTest {
+			assignments: vec![(ValidatorIndex(1), 20), (ValidatorIndex(2), 20), (ValidatorIndex(3), 20)],
+			approvals: vec![],
+			clock_drift: 10,
+			no_show_duration: 10,
+			drifted_tick_now: 20,
+			exp_no_shows: 3,
+			exp_next_no_show: None,
+		})
+	}
+
+	#[test]
+	fn count_no_shows_three_approving_validators() {
+		test_count_no_shows(NoShowTest {
+			assignments: vec![(ValidatorIndex(1), 20), (ValidatorIndex(2), 20), (ValidatorIndex(3), 20)],
+			approvals: vec![1, 2, 3],
+			clock_drift: 10,
+			no_show_duration: 10,
+			drifted_tick_now: 20,
+			exp_no_shows: 0,
+			exp_next_no_show: None,
+		})
+	}
+
+	#[test]
+	fn count_no_shows_earliest_possible_next_no_show_is_clock_drift_plus_no_show_duration() {
+		test_count_no_shows(NoShowTest {
+			assignments: vec![(ValidatorIndex(1), 0)],
+			approvals: vec![],
+			clock_drift: 10,
+			no_show_duration: 20,
+			drifted_tick_now: 0,
+			exp_no_shows: 0,
+			exp_next_no_show: Some(30),
+		})
+	}
+
+	#[test]
+	fn count_no_shows_assignment_tick_equal_to_clock_drift_yields_earliest_possible_next_no_show() {
+		test_count_no_shows(NoShowTest {
+			assignments: vec![(ValidatorIndex(1), 10)],
+			approvals: vec![],
+			clock_drift: 10,
+			no_show_duration: 20,
+			drifted_tick_now: 0,
+			exp_no_shows: 0,
+			exp_next_no_show: Some(30),
+		})
+	}
+
+	#[test]
+	fn count_no_shows_validator_index_out_of_approvals_range_is_no_show() {
+		test_count_no_shows(NoShowTest {
+			assignments: vec![(ValidatorIndex(1000), 20)],
+			approvals: vec![],
+			clock_drift: 10,
+			no_show_duration: 10,
+			drifted_tick_now: 20,
+			exp_no_shows: 1, // TODO(ladi): should this be counted or ignored?
```

Comment:
Should validator indexes past the length of the approvals vector be ignored or handled?

Reply:
Should be ignored; Garbage-in-Garbage-out